hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientCacheProxy.java

```diff
@@ -321,14 +327,142 @@ public V getAndPut(K key, V value, ExpiryPolicy expiryPolicy) {
+        } catch (Exception e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+    }
+
+    private List<Map.Entry<Data, Data>>[] groupDataToPartitions(Map<? extends K, ? extends V> map,
+                                                                ClientPartitionService partitionService,
+                                                                int partitionCount) {
+        List<Map.Entry<Data, Data>>[] entriesPerPartition = new List[partitionCount];
+        SerializationService serializationService = clientContext.getSerializationService();
+
         for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
-            put(entry.getKey(), entry.getValue(), expiryPolicy);
+            K key = entry.getKey();
+            V value = entry.getValue();
+            validateNotNull(key, value);
+
+            Data keyData = serializationService.toData(key);
+            Data valueData = serializationService.toData(value);
+
+            int partitionId = partitionService.getPartitionId(keyData);
+            List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
+            if (entries == null) {
+                entries = new ArrayList<Map.Entry<Data, Data>>();
+                entriesPerPartition[partitionId] = entries;
+            }
+
+            entries.add(new AbstractMap.SimpleImmutableEntry<Data, Data>(keyData, valueData));
+        }
+
+        return entriesPerPartition;
+    }
+
+    private static final class FutureEntryCountTuple {
+
+        private Future future;
+        private int entryCount;
+
+        private FutureEntryCountTuple(Future future, int entryCount) {
+            this.future = future;
+            this.entryCount = entryCount;
         }
+
     }
 
+    private void putToAllPartitionsAndWaitForCompletion(List<Map.Entry<Data, Data>>[] entriesPerPartition,
+                                                        ExpiryPolicy expiryPolicy, long start)
+            throws ExecutionException, InterruptedException {
+        Data expiryPolicyData = toData(expiryPolicy);
+        List<FutureEntryCountTuple> futureEntryCountTuples =
+                new ArrayList<FutureEntryCountTuple>(entriesPerPartition.length);
+        for (int partitionId = 0; partitionId < entriesPerPartition.length; partitionId++) {
+            List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
+            if (entries != null) {
+                int completionId = nextCompletionId();
+                // TODO If there is a single entry, we could make use of a put operation since that is a bit cheaper
+                ClientMessage request =
+                        CachePutAllCodec.encodeRequest(nameWithPrefix, entries, expiryPolicyData, partitionId);
+                Future f = invoke(request, partitionId, completionId);
+                futureEntryCountTuples.add(new FutureEntryCountTuple(f, entries.size()));
+                if (nearCache != null) {
```

Comment:
Yeah, that was my main point. We can update near cache after future.get() returns. I don't know if this may be necessary but we can also remove the corresponding entries from near cache if future.get() throws an exception. Then, there won't be a inconsistency case in near cache.